File: mobile/mail/lib/mail_view_preferences.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

const _standardFolders = {
  'inbox',
  'sent',
  'drafts',
  'archive',
  'spam',
  'trash',
  'all',
  'unified',
  'scheduled',
};

class MailViewPreferences {
  MailViewPreferences._();

  static String _key(String email) =>
      'cloudity.mail.view.v1:${email.trim().toLowerCase()}';

  static String _parseFolder(String? raw) {
    final f = (raw ?? '').trim();
    if (f.isEmpty) return 'inbox';
    if (_standardFolders.contains(f)) return f;
    if (f.length <= 512) return f;
    return 'inbox';
  }

  static Future<({int? accountId, String folder})> load(String email) async {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(_key(email));
    if (raw == null || raw.isEmpty) {
      return (accountId: null, folder: 'inbox');
    }
    try {
      final map = jsonDecode(raw) as Map<String, dynamic>;
      final idRaw = map['accountId'];
      final accountId = idRaw is int
          ? idRaw
          : idRaw is num
              ? idRaw.toInt()
              : int.tryParse(idRaw?.toString() ?? '');
      final folder = _parseFolder(map['folder']?.toString());
      return (
        accountId: accountId != null && accountId > 0 ? accountId : null,
        folder: folder,
      );
    } catch (_) {
      return (accountId: null, folder: 'inbox');
    }
  }

  static Future<void> save({
    required String email,
    int? accountId,
    required String folder,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      _key(email),
      jsonEncode({
        'accountId': accountId,
        'folder': _parseFolder(folder),
      }),
    );
  }
}

File: frontend/apps/cloudity-web/src/pages/app/mail/mailViewPreferences.ts
import type { MailFolderId } from '../../../api'

export type MailViewState = {
  accountId: number | null
  folder: MailFolderId
}

const LEGACY_ACCOUNT_KEY = 'cloudity_mail_selected_account_id'

const STANDARD_FOLDERS = new Set<MailFolderId>([
  'inbox',
  'sent',
  'drafts',
  'archive',
  'spam',
  'trash',
  'all',
  'unified',
  'scheduled',
])

function scopedKey(tenantId: number | null | undefined, email: string | null | undefined): string {
  const t = tenantId ?? 0
  const e = (email ?? '').trim().toLowerCase()
  return `cloudity.mail.view.v1:${t}:${e}`
}

function parseFolder(raw: unknown): MailFolderId {
  if (typeof raw !== 'string' || !raw.trim()) return 'inbox'
  const f = raw.trim() as MailFolderId
  if (STANDARD_FOLDERS.has(f)) return f
  // Dossier IMAP personnalisé (chemin libre)
  if (f.length > 0 && f.length <= 512) return f
  return 'inbox'
}

function migrateLegacyAccountId(): number | null {
  try {
    const raw = localStorage.getItem(LEGACY_ACCOUNT_KEY)
    if (!raw) return null
    const n = Number.parseInt(raw, 10)
    return Number.isFinite(n) && n > 0 ? n : null
  } catch {
    return null
  }
}

export function loadMailViewState(
  tenantId: number | null | undefined,
  email: string | null | undefined
): MailViewState {
  const fallback: MailViewState = {
    accountId: migrateLegacyAccountId(),
    folder: 'inbox',
  }
  try {
    const raw = localStorage.getItem(scopedKey(tenantId, email))
    if (!raw) return fallback
    const parsed = JSON.parse(raw) as Partial<MailViewState>
    const accountId =
      typeof parsed.accountId === 'number' && parsed.accountId > 0 ? parsed.accountId : fallback.accountId
    return {
      accountId,
      folder: parseFolder(parsed.folder),
    }
  } catch {
    return fallback
  }
}

export function saveMailViewState(
  tenantId: number | null | undefined,
  email: string | null | undefined,
  state: MailViewState
): void {
  try {
    localStorage.setItem(
      scopedKey(tenantId, email),
      JSON.stringify({
        accountId: state.accountId,
        folder: state.folder,
      })
    )
  } catch {
    /* quota / mode privé */
  }
}

/** @deprecated Préférer loadMailViewState — conservé pour tests / compat. */
export function getSavedMailSelectedAccountId(): number | null {
  return migrateLegacyAccountId()
}

/** @deprecated Préférer saveMailViewState */
export function saveMailSelectedAccountId(accountId: number | null): void {
  try {
    if (accountId == null) localStorage.removeItem(LEGACY_ACCOUNT_KEY)
    else localStorage.setItem(LEGACY_ACCOUNT_KEY, String(accountId))
  } catch {
    /* ignore */
  }
}
